File: go/quoter.go
package afi

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/http"
	"strings"
	"time"

	"github.com/ethereum/go-ethereum/common"
)

// doHTTPRequest centralises the typed-error mapping for HTTP responses.
// Returns:
//   - *NetworkError    transport / dial / decode failure
//   - *BadRequestError  4xx (caller payload bad)
//   - *ServerError      5xx (service down — retry-safe)
//   - the raw body on 2xx success
func doHTTPRequest(ctx context.Context, method, url string, payload []byte) ([]byte, error) {
	req, err := http.NewRequestWithContext(ctx, method, url, bytes.NewReader(payload))
	if err != nil {
		return nil, &NetworkError{Err: fmt.Errorf("build request: %w", err)}
	}
	if payload != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	httpClient := &http.Client{Timeout: 30 * time.Second}
	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, &NetworkError{Err: err}
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	switch {
	case resp.StatusCode >= 200 && resp.StatusCode < 300:
		return body, nil
	case resp.StatusCode >= 500:
		return nil, &ServerError{Status: resp.StatusCode, Body: string(body)}
	default:
		// 3xx is not expected from these JSON APIs — bucket with 4xx so the caller surfaces it.
		return nil, &BadRequestError{Status: resp.StatusCode, Body: string(body)}
	}
}

type quoterRequest struct {
	Network     string       `json:"network"`
	TokenIn     string       `json:"tokenIn"`
	TokenOut    string       `json:"tokenOut"`
	AmountIn    string       `json:"amountIn"`
	Slippage    float64      `json:"slippage"`
	MaxHops     int          `json:"maxHops"`
	Show        bool         `json:"show"`
	PriceBase   string       `json:"priceBase,omitempty"`
	Dexs        []string     `json:"dexs,omitempty"`
	BlockNumber string       `json:"blockNumber,omitempty"`
	RpcUrls     []RpcUrlInfo `json:"rpcUrls,omitempty"`
}

type quoterHop struct {
	TokenIn       string  `json:"tokenIn"`
	TokenOut      string  `json:"tokenOut"`
	AmountIn      string  `json:"amountIn"`
	AmountOut     string  `json:"amountOut"`
	MinOut        string  `json:"minOut"`
	AmountInRaw   string  `json:"amountInRaw"`
	AmountOutRaw  string  `json:"amountOutRaw"`
	MinOutRaw     string  `json:"minOutRaw"`
	TokenInPrice  string  `json:"tokenInPrice"`
	TokenOutPrice string  `json:"tokenOutPrice"`
	Slippage      float64 `json:"slippage"`
	Type          string  `json:"type"`
	Kind          string  `json:"kind"`
	RouteID       int     `json:"routeId"`
	Weight        float64 `json:"weight"`
}

type quoterResponseData struct {
	TokenIn           string      `json:"tokenIn"`
	TokenOut          string      `json:"tokenOut"`
	AmountIn          string      `json:"amountIn"`
	AmountOut         string      `json:"amountOut"`
	MinOut            string      `json:"minOut"`
	AmountInRaw       string      `json:"amountInRaw"`
	AmountOutRaw      string      `json:"amountOutRaw"`
	MinOutRaw         string      `json:"minOutRaw"`
	Steps             string      `json:"steps"`
	Slippage          float64     `json:"slippage"`
	Path              []string    `json:"path"`
	Hops              []quoterHop `json:"hops"`
	TokenInPrice      string      `json:"tokenInPrice"`
	TokenOutPrice     string      `json:"tokenOutPrice"`
	TokenInBasePrice  string      `json:"tokenInBasePrice"`
	TokenOutBasePrice string      `json:"tokenOutBasePrice"`
}

type quoterResponse struct {
	Status  string `json:"status"`
	Message string `json:"message"`
	// Data is an object on success and may be a string on error.
	Data json.RawMessage `json:"data"`
}

func formatAmount(amount *big.Int, decimals uint8) string {
	divisor := new(big.Int).Exp(big.NewInt(10), big.NewInt(int64(decimals)), nil)
	whole := new(big.Int).Div(amount, divisor)
	frac := new(big.Int).Mod(amount, divisor)

	if frac.Sign() == 0 {
		return whole.String()
	}

	fracStr := frac.String()
	for len(fracStr) < int(decimals) {
		fracStr = "0" + fracStr
	}
	fracStr = strings.TrimRight(fracStr, "0")
	return whole.String() + "." + fracStr
}

// fetchQuote is called by the client, passing the resolved options and quoter URL.
func fetchQuote(ctx context.Context, opts *quoteOptions, feeBps uint16, quoterURL string) (*Quote, error) {
	return fetchQuoteFrom(ctx, opts, feeBps, quoterURL)
}

// fetchQuoteFrom is the testable version — same params but quoterURL is injectable.
func fetchQuoteFrom(ctx context.Context, opts *quoteOptions, feeBps uint16, quoterURL string) (*Quote, error) {
	body := quoterRequest{
		Network:  string(opts.network),
		TokenIn:  opts.tokenIn,
		TokenOut: opts.tokenOut,
		AmountIn: opts.amountIn,
		Slippage: opts.slippage,
		MaxHops:  opts.maxHops,
		Show:     true,
	}
	if opts.priceBase != "" {
		body.PriceBase = opts.priceBase
	}
	if len(opts.dexs) > 0 {
		dexStrs := make([]string, len(opts.dexs))
		for i, d := range opts.dexs {
			dexStrs[i] = string(d)
		}
		body.Dexs = dexStrs
	}
	if opts.blockNumber != "" {
		body.BlockNumber = opts.blockNumber
	}
	if len(opts.rpcUrls) > 0 {
		body.RpcUrls = opts.rpcUrls
	}

	// json.Marshal cannot fail for a struct with only basic scalar fields
	payload, _ := json.Marshal(body)

	respBody, err := doHTTPRequest(ctx, http.MethodPost, quoterURL, payload)
	if err != nil {
		// Surface the typed error directly — callers can still match QUOTE_FAILED via
		// errors.As but get the precise BadRequest / Server / Network bucket too.
		return nil, err
	}

	var result quoterResponse
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, ErrQuote(fmt.Sprintf("decode: %v", err))
	}

	if result.Status != "success" {
		msg := result.Message
		if msg == "" {
			// data may carry the error as a plain string
			var dataStr string
			if json.Unmarshal(result.Data, &dataStr) == nil && dataStr != "" {
				msg = dataStr
			}
		}
		if msg == "" {
			msg = "unknown error from quoter"
		}
		return nil, ErrQuote(msg)
	}

	var d quoterResponseData
	if err := json.Unmarshal(result.Data, &d); err != nil {
		return nil, ErrQuote(fmt.Sprintf("decode data: %v", err))
	}

	if d.MinOutRaw == "" || d.MinOutRaw == "0" {
		return nil, ErrQuote("received zero minOut — rejected for safety")
	}

	amountInWei := new(big.Int)
	amountInWei.SetString(d.AmountInRaw, 10)

	amountOutWei := new(big.Int)
	amountOutWei.SetString(d.AmountOutRaw, 10)

	minOutWei := new(big.Int)
	minOutWei.SetString(d.MinOutRaw, 10)

	steps := common.FromHex(d.Steps)

	path := make([]common.Address, len(d.Path))
	for i, p := range d.Path {
		path[i] = common.HexToAddress(p)
	}

	hops := make([]Hop, len(d.Hops))
	for i, h := range d.Hops {
		amountInWeiHop := new(big.Int)
		amountInWeiHop.SetString(h.AmountInRaw, 10)
		amountOutWeiHop := new(big.Int)
		amountOutWeiHop.SetString(h.AmountOutRaw, 10)
		minOutWeiHop := new(big.Int)
		minOutWeiHop.SetString(h.MinOutRaw, 10)
		hops[i] = Hop{
			TokenIn:       common.HexToAddress(h.TokenIn),
			TokenOut:      common.HexToAddress(h.TokenOut),
			AmountIn:      h.AmountIn,
			AmountOut:     h.AmountOut,
			MinOut:        h.MinOut,
			AmountInWei:   amountInWeiHop,
			AmountOutWei:  amountOutWeiHop,
			MinOutWei:     minOutWeiHop,
			TokenInPrice:  h.TokenInPrice,
			TokenOutPrice: h.TokenOutPrice,
			Slippage:      h.Slippage,
			Type:          h.Type,
			Kind:          h.Kind,
			RouteID:       h.RouteID,
			Weight:        h.Weight,
		}
	}

	return &Quote{
		TokenIn:           common.HexToAddress(d.TokenIn),
		TokenOut:          common.HexToAddress(d.TokenOut),
		AmountIn:          d.AmountIn,
		AmountOut:         d.AmountOut,
		MinOut:            d.MinOut,
		AmountInWei:       amountInWei,
		AmountOutWei:      amountOutWei,
		MinOutWei:         minOutWei,
		Steps:             steps,
		Path:              path,
		Hops:              hops,
		Slippage:          d.Slippage,
		FeeBps:            feeBps,
		TokenInPrice:      d.TokenInPrice,
		TokenOutPrice:     d.TokenOutPrice,
		TokenInBasePrice:  d.TokenInBasePrice,
		TokenOutBasePrice: d.TokenOutBasePrice,
		CreatedAt:         timeNowMS(),
		Network:           opts.network,
		MaxHops:           opts.maxHops,
		PriceBase:         opts.priceBase,
		Dexs:              opts.dexs,
	}, nil
}

// ─── Generic JSON endpoint helpers ───────────────────────────────────────────

// commandRequest wraps a /command payload with the action discriminator.
type commandRequest struct {
	Action string                 `json:"action"`
	Body   map[string]interface{} `json:"-"`
}

// marshal merges Action + Body into a single JSON object.
func (cr commandRequest) marshal() ([]byte, error) {
	m := make(map[string]interface{}, len(cr.Body)+1)
	for k, v := range cr.Body {
		m[k] = v
	}
	m["action"] = cr.Action
	return json.Marshal(m)
}

// postCommand POSTs to /command with the given action + payload, decoding the
// service envelope ({status, data, time}) and unmarshalling `data` into out.
func (c *Client) postCommand(ctx context.Context, action string, body map[string]interface{}, out interface{}) error {
	payload, err := commandRequest{Action: action, Body: body}.marshal()
	if err != nil {
		return ErrQuote(fmt.Sprintf("marshal: %v", err))
	}
	raw, err := doHTTPRequest(ctx, http.MethodPost, c.apiURL+"/command", payload)
	if err != nil {
		return err
	}
	return decodeEnvelope(raw, out)
}

// decodeEnvelope unwraps the standard afi-rpc response envelope
// ({status, data, time}): on a non-success status it returns the error message,
// otherwise it unmarshals the `data` field into out (a no-op when out is nil or
// data is empty).
func decodeEnvelope(raw []byte, out interface{}) error {
	var env struct {
		Status string          `json:"status"`
		Data   json.RawMessage `json:"data"`
	}
	if err := json.Unmarshal(raw, &env); err != nil {
		return ErrQuote(fmt.Sprintf("decode envelope: %v", err))
	}
	if env.Status != "" && env.Status != "success" {
		var msg string
		_ = json.Unmarshal(env.Data, &msg)
		if msg == "" {
			msg = env.Status
		}
		return ErrQuote(msg)
	}
	if out == nil || len(env.Data) == 0 {
		return nil
	}
	if err := json.Unmarshal(env.Data, out); err != nil {
		return ErrQuote(fmt.Sprintf("decode data: %v", err))
	}
	return nil
}

// postForData POSTs req to url and decodes the response envelope's data into out.
func postForData(ctx context.Context, url string, req, out interface{}) error {
	payload, err := json.Marshal(req)
	if err != nil {
		return ErrQuote(fmt.Sprintf("marshal: %v", err))
	}
	raw, err := doHTTPRequest(ctx, http.MethodPost, url, payload)
	if err != nil {
		return err
	}
	return decodeEnvelope(raw, out)
}

// ───────────────── HTTP endpoint wrappers ─────────────────

// ArbitrageRequest is the body of POST /arbitrage. Use raw fields (tokenIn,
// tokenOut, amountIn, network) — the RPC service interprets the shape. For a
// self-funded cycle set tokenIn == tokenOut.
type ArbitrageRequest map[string]interface{}

// FindArbitrage hits POST /arbitrage and returns the candidate routes, decoding
// the service envelope ({status, data, time}). Each RouteQuote is an executable
// single-DEX route — feed the most profitable one to QuoteFromRoute.
func (c *Client) FindArbitrage(ctx context.Context, req ArbitrageRequest) ([]RouteQuote, error) {
	var routes []RouteQuote
	if err := postForData(ctx, c.apiURL+"/arbitrage", req, &routes); err != nil {
		return nil, err
	}
	return routes, nil
}

// PathRequest is the body of POST /command action="path".
type PathRequest map[string]interface{}

// FindPath hits POST /command action="path" and returns the priced multi-hop
// route for an explicit token path.
func (c *Client) FindPath(ctx context.Context, req PathRequest) (*PathQuote, error) {
	var resp PathQuote
	if err := c.postCommand(ctx, "path", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// RoutesRequest is the body of POST /command action="routes".
type RoutesRequest map[string]interface{}

// GetRoutes hits POST /command action="routes" and returns the candidate token
// paths between tokenIn and tokenOut.
func (c *Client) GetRoutes(ctx context.Context, req RoutesRequest) ([]Route, error) {
	var routes []Route
	if err := c.postCommand(ctx, "routes", req, &routes); err != nil {
		return nil, err
	}
	return routes, nil
}

// LiquidationCandidatesRequest is the body of POST /aave.
type LiquidationCandidatesRequest map[string]interface{}

// GetLiquidationCandidates hits POST /aave and returns open Aave positions
// eligible for liquidation.
func (c *Client) GetLiquidationCandidates(ctx context.Context, req LiquidationCandidatesRequest) ([]AavePosition, error) {
	var positions []AavePosition
	if err := postForData(ctx, c.apiURL+"/aave", req, &positions); err != nil {
		return nil, err
	}
	return positions, nil
}

// LiquidateRequest is the body of POST /liquidation-call.
type LiquidateRequest map[string]interface{}

// Liquidate hits POST /liquidation-call and returns the executable route that
// repays the debt and swaps the seized collateral back.
func (c *Client) Liquidate(ctx context.Context, req LiquidateRequest) (*LiquidationResult, error) {
	var resp LiquidationResult
	if err := postForData(ctx, c.apiURL+"/liquidation-call", req, &resp); err != nil {
		return nil, err
	}
	return &resp, nil
}

// PriceQuoteRequest is the body of POST /command action="price".
type PriceQuoteRequest map[string]interface{}

// PriceQuote hits POST /command action="price" and returns the per-DEX quotes
// for the pair (same shape as FindArbitrage).
func (c *Client) PriceQuote(ctx context.Context, req PriceQuoteRequest) ([]RouteQuote, error) {
	var quotes []RouteQuote
	if err := c.postCommand(ctx, "price", req, &quotes); err != nil {
		return nil, err
	}
	return quotes, nil
}

// DexQuoteRequest is the body of POST /command action=dex.
type DexQuoteRequest map[string]interface{}

// QuoteDex hits POST /command action=dex (e.g. "uniV3", "aerodrome") and returns
// that DEX's quotes for the pair.
func (c *Client) QuoteDex(ctx context.Context, dex string, req DexQuoteRequest) ([]RouteQuote, error) {
	var quotes []RouteQuote
	if err := c.postCommand(ctx, dex, req, &quotes); err != nil {
		return nil, err
	}
	return quotes, nil
}

File: nodejs/src/quoter.ts
import { BadRequestError, NetworkError, QuoteError, ServerError } from "./errors.js"
import { encodeSteps } from "./encode-steps.js"
import type { Address, Dex, Hex, Hop, Network, Quote, RpcUrlInfo } from "./types.js"

export interface QuoteRequest {
  tokenIn: Address
  tokenOut: Address
  amountIn: string
  slippage: number
  maxHops: number
  network: Network
  priceBase?: string
  dexs?: Dex[]
  blockNumber?: string | number
  rpcUrls?: RpcUrlInfo[]
}

interface QuoterHop {
  tokenIn: string
  tokenOut: string
  amountIn: string
  amountOut: string
  minOut: string
  amountInRaw: string
  amountOutRaw: string
  minOutRaw: string
  tokenInPrice: string
  tokenOutPrice: string
  slippage: number
  type: string
  kind: string
  routeId: number
  weight: number
}

interface QuoterResponseData {
  tokenIn: string
  tokenOut: string
  amountIn: string
  amountOut: string
  minOut: string
  amountInRaw: string
  amountOutRaw: string
  minOutRaw: string
  steps: string
  slippage: number
  path: string[]
  hops: QuoterHop[]
  tokenInPrice: string
  tokenOutPrice: string
  tokenInBasePrice?: string
  tokenOutBasePrice?: string
}

interface QuoterApiResponse {
  status: "success" | "error"
  message?: string
  data?: QuoterResponseData | string
}

export async function fetchQuote(
  params: QuoteRequest,
  feeBps: number,
  quoterUrl: string,
): Promise<Quote> {
  const body: Record<string, unknown> = {
    network:  params.network,
    tokenIn:  params.tokenIn,
    tokenOut: params.tokenOut,
    amountIn: params.amountIn,
    slippage: params.slippage,
    maxHops:  params.maxHops,
    show:     true,
  }

  if (params.rpcUrls && params.rpcUrls.length > 0) {
    body.rpcUrls = params.rpcUrls
  }
  if (params.priceBase !== undefined) {
    body.priceBase = params.priceBase
  }
  if (params.dexs && params.dexs.length > 0) {
    body.dexs = params.dexs
  }
  if (params.blockNumber !== undefined) {
    body.blockNumber = params.blockNumber
  }

  let res: Response
  try {
    res = await fetch(quoterUrl, {
      method:  "POST",
      headers: { "Content-Type": "application/json" },
      body:    JSON.stringify(body),
    })
  } catch (e) {
    throw new NetworkError(`network error: ${(e as Error).message}`)
  }

  if (!res.ok) {
    const msg = `HTTP ${res.status}`
    if (res.status >= 500) throw new ServerError(msg, res.status)
    if (res.status >= 400) throw new BadRequestError(msg, res.status)
    throw new QuoteError(msg)
  }

  const json = (await res.json()) as QuoterApiResponse

  if (json.status !== "success" || !json.data || typeof json.data === "string") {
    const msg =
      json.message ??
      (typeof json.data === "string" ? json.data : undefined) ??
      "unknown error from quoter"
    throw new QuoteError(msg)
  }

  const d = json.data

  if (!d.minOutRaw || d.minOutRaw === "0") {
    throw new QuoteError("received zero minOut — rejected for safety")
  }

  const hops: Hop[] = (d.hops ?? []).map((h) => ({
    tokenIn:       h.tokenIn as Address,
    tokenOut:      h.tokenOut as Address,
    amountIn:      h.amountIn,
    amountOut:     h.amountOut,
    minOut:        h.minOut,
    amountInWei:   BigInt(h.amountInRaw),
    amountOutWei:  BigInt(h.amountOutRaw),
    minOutWei:     BigInt(h.minOutRaw),
    tokenInPrice:  h.tokenInPrice,
    tokenOutPrice: h.tokenOutPrice,
    slippage:      h.slippage,
    type:          h.type,
    kind:          h.kind,
    routeId:       h.routeId,
    weight:        h.weight,
  }))

  const quote: Quote = {
    tokenIn:       d.tokenIn as Address,
    tokenOut:      d.tokenOut as Address,
    amountIn:      d.amountIn,
    amountOut:     d.amountOut,
    minOut:        d.minOut,
    amountInWei:   BigInt(d.amountInRaw),
    amountOutWei:  BigInt(d.amountOutRaw),
    minOutWei:     BigInt(d.minOutRaw),
    steps:         d.steps as Hex,
    path:          d.path as Address[],
    hops,
    slippage:      d.slippage,
    feeBps,
    tokenInPrice:  d.tokenInPrice,
    tokenOutPrice: d.tokenOutPrice,
    createdAt:     Date.now(),
    network:       params.network,
    maxHops:       params.maxHops,
    priceBase:     params.priceBase,
    dexs:          params.dexs,
  }

  if (d.tokenInBasePrice !== undefined)  quote.tokenInBasePrice  = d.tokenInBasePrice
  if (d.tokenOutBasePrice !== undefined) quote.tokenOutBasePrice = d.tokenOutBasePrice

  return quote
}

// ─── Generic afi-rpc HTTP helpers ────────────────────────────────────────────

/**
 * Base URL passed to all helpers in this section is the API root (e.g.
 * `https://rpc.afi.run`) — NOT a specific endpoint. Each helper appends its
 * own path.
 */

interface GenericApiResponse<T = unknown> {
  status: "success" | "error"
  message?: string
  data?: T | string
}

async function postJson<T>(url: string, body: Record<string, unknown>): Promise<T> {
  let res: Response
  try {
    res = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    })
  } catch (e) {
    throw new NetworkError(`network error: ${(e as Error).message}`)
  }

  if (!res.ok) {
    const msg = `HTTP ${res.status}`
    if (res.status >= 500) throw new ServerError(msg, res.status)
    if (res.status >= 400) throw new BadRequestError(msg, res.status)
    throw new QuoteError(msg)
  }

  const json = (await res.json()) as GenericApiResponse<T>
  if (json.status !== "success" || json.data === undefined || typeof json.data === "string") {
    const msg =
      json.message ??
      (typeof json.data === "string" ? json.data : undefined) ??
      "unknown error from afi-rpc"
    throw new QuoteError(msg)
  }
  return json.data as T
}

// ─── /arbitrage ──────────────────────────────────────────────────────────────

export interface ArbitrageRequest {
  network: Network
  tokenIn: Address
  /** Cycle end token — set equal to `tokenIn` for a self-funded cycle. */
  tokenOut: Address
  amountIn: string
  /** Optional list of intermediate tokens to consider. */
  tokens?: Address[]
  maxHops?: number
  slippage?: number
  dexs?: Dex[]
  rpcUrls?: RpcUrlInfo[]
}

/**
 * RouteQuote is one route returned by /arbitrage, /command "price", or a per-DEX
 * quote — a single-DEX quote for tokenIn → tokenOut with the encoded hop ready
 * to run through Afi.swap. Feed the best one to quoteFromRoute.
 */
export interface RouteQuote {
  network: string
  type?: string
  kind?: string
  tokenIn: Address
  tokenOut: Address
  tokenInPrice?: string
  tokenOutPrice?: string
  amountIn: string
  amountInRaw: string
  amountOut: string
  amountOutRaw: string
  minOut?: string
  minOutRaw: string
  slippage?: number
  weight?: number
  blockNumber?: string
  fee?: number
  routeId: number
  /** Opaque per-hop calldata for the DEX adapter, hex-encoded. */
  stepData: Hex
}

/** routeProfit returns amountOutRaw − amountInRaw, or null if either is unparseable. */
export function routeProfit(r: RouteQuote): bigint | null {
  try {
    return BigInt(r.amountOutRaw) - BigInt(r.amountInRaw)
  } catch {
    return null
  }
}

/**
 * quoteFromRoute hydrates an executable Quote from a RouteQuote: it wraps the
 * route's single hop ({routeId, stepData}) into Afi.swap params via encodeSteps
 * and copies the token/amount fields. The result can go straight to
 * client.executeSwap / client.simulate.
 *
 * `minOutOverride` replaces the route's minOutRaw — handy for a self-funded
 * cycle that should floor output at principal + threshold and revert otherwise.
 */
export function quoteFromRoute(r: RouteQuote, minOutOverride?: bigint): Quote {
  if (r.routeId < 0 || r.routeId > 0xffff) {
    throw new Error(`quoteFromRoute: routeId ${r.routeId} out of uint16 range`)
  }
  const steps = encodeSteps([{ id: r.routeId, data: (r.stepData ?? "0x") as Hex }])
  return {
    tokenIn: r.tokenIn,
    tokenOut: r.tokenOut,
    amountIn: r.amountIn ?? "",
    amountOut: r.amountOut ?? "",
    minOut: r.minOut ?? "",
    amountInWei: BigInt(r.amountInRaw),
    amountOutWei: r.amountOutRaw ? BigInt(r.amountOutRaw) : 0n,
    minOutWei: minOutOverride ?? BigInt(r.minOutRaw),
    steps,
    path: [r.tokenIn, r.tokenOut],
    hops: [],
    slippage: r.slippage ?? 0,
    feeBps: 0,
    tokenInPrice: r.tokenInPrice ?? "",
    tokenOutPrice: r.tokenOutPrice ?? "",
    createdAt: Date.now(),
    network: r.network as Network,
    maxHops: 1,
  }
}

/**
 * findArbitrage hits POST /arbitrage and returns the candidate routes. Each
 * RouteQuote is an executable single-DEX route — feed the most profitable one to
 * quoteFromRoute. For a self-funded cycle set tokenIn === tokenOut.
 */
export async function findArbitrage(
  apiBaseUrl: string,
  req: ArbitrageRequest,
): Promise<RouteQuote[]> {
  return postJson<RouteQuote[]>(`${apiBaseUrl}/arbitrage`, { ...req })
}

// ─── /command  (dispatch table for action-based endpoints) ──────────────────

export interface CommandRequestBase {
  network: Network
  rpcUrls?: RpcUrlInfo[]
}

export interface PathRequest extends CommandRequestBase {
  tokenIn: Address
  tokenOut: Address
  amountIn?: string
  maxHops?: number
}

/** PathQuote is the priced multi-hop route returned by findPath. */
export interface PathQuote {
  network: string
  path: Address[]
  tokenIn: Address
  tokenOut: Address
  amountIn: string
  amountInRaw: string
  amountOut: string
  amountOutRaw: string
  minOut: string
  minOutRaw: string
  tokenInPrice?: string
  tokenOutPrice?: string
  blockNumber?: string
  slippage?: number
  /** Combined hex-encoded params ready for Afi.swap. */
  steps: Hex
  hops: RouteQuote[]
}

export async function findPath(apiBaseUrl: string, req: PathRequest): Promise<PathQuote> {
  return postJson<PathQuote>(`${apiBaseUrl}/command`, { action: "path", ...req })
}

export interface RoutesRequest extends CommandRequestBase {
  tokenIn?: Address
  tokenOut?: Address
  maxHops?: number
}

/** Route is one candidate token path returned by getRoutes. */
export interface Route {
  path: Address[]
}

export async function getRoutes(apiBaseUrl: string, req: RoutesRequest): Promise<Route[]> {
  return postJson<Route[]>(`${apiBaseUrl}/command`, { action: "routes", ...req })
}

export interface PriceQuoteRequest extends CommandRequestBase {
  tokenIn: Address
  tokenOut: Address
  amountIn: string
}

/** priceQuote returns the per-DEX quotes for the pair (same shape as findArbitrage). */
export async function priceQuote(
  apiBaseUrl: string,
  req: PriceQuoteRequest,
): Promise<RouteQuote[]> {
  return postJson<RouteQuote[]>(`${apiBaseUrl}/command`, { action: "price", ...req })
}

/** Per-DEX quote action names dispatched via `/command`. */
export type DexAction =
  | "uniV3"
  | "cakeV3"
  | "uniV4"
  | "aerodrome"
  | "balancer"
  | "fluid"
  | "curve128"
  | "curve256"

export interface DexQuoteRequest extends CommandRequestBase {
  tokenIn: Address
  tokenOut: Address
  amountIn: string
}

export async function quoteDex(
  apiBaseUrl: string,
  dex: DexAction,
  req: DexQuoteRequest,
): Promise<RouteQuote[]> {
  return postJson<RouteQuote[]>(`${apiBaseUrl}/command`, { action: dex, ...req })
}

// ─── /aave + /liquidation-call ───────────────────────────────────────────────

export interface LiquidationCandidatesRequest {
  network: Network
  skip?: number
  first?: number
  currentTotalDebt?: number
  rpcUrls?: RpcUrlInfo[]
}

export interface AaveCollateral {
  token: string
  tokenAddress: string
  aToken: string
  balanceRaw: string
  balance: string
  decimals: number
}

export interface AavePosition {
  user: string
  debtToken: string
  debtTokenAddress: string
  debtAToken: string
  decimals: number
  debtAmountRaw: string
  debtAmount: string
  collaterals: AaveCollateral[]
}

export async function getLiquidationCandidates(
  apiBaseUrl: string,
  req: LiquidationCandidatesRequest,
): Promise<AavePosition[]> {
  return postJson<AavePosition[]>(`${apiBaseUrl}/aave`, { ...req })
}

export interface LiquidateRequest {
  network: Network
  pool: Address
  user: Address
  tokenIn: Address
  tokenOut: Address
  amountIn: string
  slippage?: number
  rpcUrls?: RpcUrlInfo[]
}

/** LiquidationResult is the executable repay+swap route returned by liquidate. */
export interface LiquidationResult {
  tokenIn: Address
  tokenOut: Address
  amountIn: string
  amountOut: string
  profit: string
  blockNumber?: unknown
  slippage?: number
  /** Combined hex-encoded params ready for Afi.swap. */
  steps: Hex
  hops: RouteQuote[]
}

export async function liquidate(
  apiBaseUrl: string,
  req: LiquidateRequest,
): Promise<LiquidationResult> {
  return postJson<LiquidationResult>(`${apiBaseUrl}/liquidation-call`, { ...req })
}
